### PR TITLE
Bigquery

### PR DESCRIPTION
**Pull request by drewbanin**
```
-- profile.yml
bq:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: oauth
      project: 'project_name'
      schema: "your_schema" # dataset
      threads: 1
```

### CODE REVIEW

File: dbt/adapters/bigquery.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+        if state is None:
+            state = 'SUCCESS'
+
+        return "{} {}".format(state, cursor.rowcount)
+
+    @classmethod
+    def get_bigquery_credentials(cls, config):
+        method = config.get('method')
+        Creds = google.oauth2.service_account.Credentials
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: don't upcase variable names

File: dbt/adapters/bigquery.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+        if method == 'oauth':
+            return None
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: what happens if they use the oauth method?

File: dbt/adapters/bigquery.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        elif method == 'service-account':
+            keyfile = config.get('keyfile')
+            return Creds.from_service_account_file(keyfile)
+
+        elif method == 'service-account-json':
+            details = config.get('config')
+            return Creds.from_service_account_info(details)
+
+        error = ('Bad `method` in profile: "{}". '
+                 'Should be "oauth" or "service-account"'.format(method))
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: or service-account-json?

Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: also can we put this validation into a voluptuous schema for the bigquery profile?

File: dbt/adapters/bigquery.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        view = dataset.table(view_name)
+        view.delete()
+
+    @classmethod
+    def rename(cls, profile, from_name, to_name, model_name=None):
+        message = 'Cannot rename bigquery relation {} to {}'.format(
+                  from_name, to_name)
+        raise dbt.exceptions.NotImplementedException(message)
+
+    # hack because of current API limitations
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: can you elaborate more? if there's a link explaining why we need this, you should add it here

File: dbt/adapters/bigquery.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+        connection_name = connection.get('name')
+
+        sql = 'select system$abort_session({})'.format(sid)
+
+        logger.debug("Cancelling query '{}' ({})".format(connection_name, sid))
+
+        _, cursor = cls.add_query(profile, sql, 'master')
+        res = cursor.fetchone()
+
+        logger.debug("Cancel query '{}': {}".format(connection_name, res))
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: what's the behavior on ctrl+c here?

if we need to just raise an exception, you should remove all this code.

also please change this and the exception for `add_query` to dbt.exceptions.NotImplementedException

File: dbt/adapters/default.py [hunk 6]
```diff
@@ -21,6 +23,33 @@
...
+        return importlib.import_module(name)
+
+    @classmethod
+    def install_requires(cls):
+        from pip import main as pip_main
+
+        for package, require in cls.requires.items():
+            logger.info("Installing {}".format(require))
+            pip_main(['install', require])
+            logger.info("Installed {} successfully!".format(require))
```
Comment (cmcarthur) [inline dbt/adapters/default.py]: is there a unit test for this? i want to be sure it works on all platforms

File: dbt/contracts/connection.py [hunk 7]
```diff
@@ -33,10 +33,19 @@
     Optional('role'): basestring,
 })
 
+bigquery_credentials_contract = Schema({
+    Required('method'): basestring,
```
Comment (cmcarthur) [inline dbt/contracts/connection.py]: `All(basestring, Any(['oauth', 'service-account', 'service-account-json']))`

File: dbt/logger.py [hunk 8]
```diff
@@ -44,6 +43,7 @@
 
 
 def make_log_dir_if_missing(log_dir):
+    import dbt.clients.system
     dbt.clients.system.make_directory(log_dir)
```
Comment (cmcarthur) [inline dbt/logger.py]: why move this import in here?

File: test/unit/test_runner.py [hunk 9]
```diff
@@ -93,16 +101,13 @@ def test__execute_model__view(self,
...
-            model,
-            existing=self.existing)
+        runner = self.get_model_runner(model)
+        runner.execute(model, existing=self.existing)
 
         dbt.adapters.postgres.PostgresAdapter.drop.assert_not_called()
 
         mock_adapter_truncate.assert_not_called()
-        mock_adapter_rename.assert_not_called()
+        mock_adapter_rename.assert_called_once()
```
Comment (cmcarthur) [inline test/unit/test_runner.py]: this makes me nervous. was this not working properly before? my understanding that view models would just be created with the final name if they did not previously exist (vs created and then renamed)

File: dbt/contracts/connection.py (hunk 7 above)
Comment (drewbanin) [inline dbt/contracts/connection.py]: Can i just do one of these:
```
Any(['oauth', 'service-account', 'service-account-json'])
```

File: dbt/adapters/bigquery.py (hunk 2 above)
Comment (drewbanin) [inline dbt/adapters/bigquery.py]: We initialize a bigquery client with the `credentials` argument if it's a service account (from a private key file or the private key raw json).

If the credentials object is `None` (as with oauth), then the bq client lib will use the oauth flow to authenticate the user. We force this flow if the user isn't authenticated by shelling out to `gcloud` [here](https://github.com/fishtown-analytics/dbt/pull/455/files#diff-a8966822c3061ac7ef67f0f6416e5cd8R135)

More info on gcp auth flows [here](https://googlecloudplatform.github.io/google-cloud-python/stable/google-cloud-auth.html#credential-discovery-precedence)

File: dbt/adapters/bigquery.py (hunk 4 above)
Comment (drewbanin) [inline dbt/adapters/bigquery.py]: Sure, it's actually not super well documented which is part of the problem. BQ defaults to the "Legacy SQL" syntax when views are created via the API. You can set a `useLegacySql = False` flag on the `Table` object, but it looks like it's presently ignored by the API.

[Relevant issue](https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3388)

The best workaround i could find was to use the `#standardSQL` header line


[Relevant issue](https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3388)

File: dbt/adapters/bigquery.py (hunk 5 above)
Comment (drewbanin) [inline dbt/adapters/bigquery.py]: @cmcarthur hadn't gotten there yet :)

Looks like we can actually issue a cancel command for running queries: https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/cancel

File: dbt/logger.py (hunk 8 above)
Comment (drewbanin) [inline dbt/logger.py]: the logger imports the system client, and the system client imports the logger. This breaks the circular import

File: test/unit/test_runner.py (hunk 9 above)
Comment (drewbanin) [inline test/unit/test_runner.py]: i'm really dissatisfied with our current drop/create/rename logic. It works ok, but that `__dbt_tmp` suffix makes a lot of coordination between dbt components unduly difficult. I think this was an attempt at simplifying how it works, but let me revisit. Good catch

File: dbt/adapters/bigquery.py (hunk 5 above)
Comment (drewbanin) [inline dbt/adapters/bigquery.py]: @cmcarthur it doesn't look like it's possible to cancel a table creation request. Jobs can be canceled though, which may be good to keep in mind for the future. This branch will log an error that looks like this on ctrl-c:

```
The bigquery adapter does not support query cancellation. Some queries may still be running!
```

**Comment by drewbanin** [comment]
@cmcarthur this is pretty much good. I did change the way views are materialized (temp table, then renamed), so the changed unit test is warranted imo.

I have a single integration test for BQ which passes on circle and appveyor.

Think this is ready for another look!

File: dbt/adapters/bigquery.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    def begin(cls, profile, name='master'):
+        pass
+
+    @classmethod
+    def commit(cls, connection):
+        pass
+
+    @classmethod
+    def get_status(cls, cursor):
+        raise Exception("Not implemented")
```
Comment (cmcarthur) [inline dbt/adapters/bigquery.py]: NotImplementedException // remove code below

Comment (drewbanin) [inline dbt/adapters/bigquery.py]: good catch :+1:

**Pull request merged**